src/js/media-controller.js

```diff
@@ -0,0 +1,309 @@
+      // receive these state updates
+      if (typeof child[stateName] !== 'undefined') {
+        child[stateName] = val;
+      }
+      propagateMediaState(child, stateName, val);
+    }
+
+    // Make sure custom els are ready
+    if (childName.includes('-') && !window.customElements.get(childName)) {
+      window.customElements.whenDefined(childName).then(setAndPropagate);
+    } else {
+      setAndPropagate();
+    }
+  });
+}
+
+class MediaController extends window.HTMLElement {
+  constructor() {
+    super();
+
+    // Set up the Shadow DOM
+    const shadow = this.attachShadow({ mode: 'open' });
+    this.shadowRoot.appendChild(template.content.cloneNode(true));
+    this.container = this.shadowRoot.getElementById('container');
+
+    // Watch for child adds/removes and update the media element if necessary
+    const mutationCallback = (mutationsList, observer) => {
+      const media = this.media;
+
+      for (let mutation of mutationsList) {
+        if (mutation.type === 'childList') {
+
+          // Media element being removed
+          mutation.removedNodes.forEach(node => {
+            // Is this a direct child media element of media-controller?
+            // TODO: This accuracy doesn't matter after moving away from media attrs.
+            // Could refactor so we can always just call 'dispose' on any removed media el.
+            if (node.slot == 'media' && mutation.target == this) {
+              // Check if this was the current media by if it was the first
+              // el with slot=media in the child list. There could be multiple.
+              let previousSibling = mutation.previousSibling && mutation.previousSibling.previousElementSibling;
+
+              // Must have been first if no prev sibling or new media
+              if (!previousSibling || !media) {
+                this.mediaUnsetCallback(node);
+              } else {
+                // Check if any prev siblings had a slot=media
+                // Should remain true otherwise
+                let wasFirst = previousSibling.slot !== 'media';
+                while ((previousSibling = previousSibling.previousSibling) !== null) {
+                  if (previousSibling.slot == 'media') wasFirst = false;
+                }
+                if (wasFirst) this.mediaUnsetCallback(node);
+              }
+            }
+          });
+
+          // Controls or media element being added
+          // No need to inject anything if media=null
+          if (media) {
+            mutation.addedNodes.forEach(node => {
+              if (node == media) {
+                // Update all controls with new media if this is the new media
+                this.mediaSetCallback(node);
+              }
+            });
+          }
+        }
+      }
+    };
+
+    const observer = new MutationObserver(mutationCallback);
+    observer.observe(this, { childList: true, subtree: true });
+
+    // Capture requests from internal control events
```

Comment:
Quick POC of handling events that bubble from child elements or dispatched by associated elements.